### PR TITLE
Add Go to cookbook

### PR DESCRIPTION
**Pull request by R-maan**

### CODE REVIEW

File: guides/cookbook.md [hunk 1]
```diff
@@ -224,6 +224,85 @@ class WriteIonText
 If Ion binary encoding is desired, use `IonBinaryWriterBuilder` (instead of `IonTextWriterBuilder`).
 </div>
 
+<div class="tabpane Go" markdown="1">
+Implementations of the [`Reader`][22] and [`Writer`][23] interfaces are responsible for reading and writing Ion data in both text and binary forms.
+
+In order to make and use a text reader, we can use `NewReaderString()`. The following example demonstrates how to read an Ion data from a string:
```
Comment (matthewh-BQ) [inline guides/cookbook.md]:
```suggestion
In order to make and use a text reader, we can use `NewReaderString()`. The following example demonstrates how to read Ion data from a string:
```

File: guides/cookbook.md [hunk 2]
```diff
@@ -224,6 +224,85 @@ class WriteIonText
...
+		currentType = reader.Type()                              // the first value in the struct is of type string
+		fmt.Println("Current type, after stepping in the struct:\t" +
+			currentType.String())          // Current type, after stepping in the struct:   string
+		fieldName := reader.FieldName()    // retrieve the current value's field name
+		value, err := reader.StringValue() // retrieve the current value's string value
+		if err != nil {
+			panic("Reading string value failed.")
+		}
+		reader.StepOut()                    // step out of the struct
+		fmt.Println(*fieldName, " ", value) // hello   world
```
Comment (matthewh-BQ) [inline guides/cookbook.md]:
```suggestion
		fmt.Println(*fieldName, " ", value) // hello world
```

File: guides/cookbook.md [hunk 3]
```diff
@@ -224,6 +224,85 @@ class WriteIonText
...
+	}
+}
+```
+Should we have a binary Ion, `NewReaderBytes()` can be used in the same fashion.
```
Comment (matthewh-BQ) [inline guides/cookbook.md]:
```suggestion
If we have a binary Ion, `NewReaderBytes()` can be used in the same fashion.
```

File: guides/cookbook.md [hunk 4]
```diff
@@ -725,6 +804,52 @@ class PrettyPrint
 ```
 </div>
 
+<div class="tabpane Go" markdown="1">
+Ion data can be pretty-printed using `NewTextWriterOpts` and pass `TextWriterPretty` option to it:
```
Comment (matthewh-BQ) [inline guides/cookbook.md]:
```suggestion
Ion data can be pretty-printed using `NewTextWriterOpts` and by passing the `TextWriterPretty` option to it:
```

File: guides/cookbook.md [hunk 5]
```diff
@@ -224,6 +224,85 @@ class WriteIonText
...
+
+import (
+	"fmt"
+	"github.com/amzn/ion-go/ion"
+)
+
+func main() {
+	reader := ion.NewReaderString("{hello:\"world\"}")
+	if reader.Next() { 											 // position the reader at the first value
+		currentType := reader.Type()                             // the first value in the reader is struct
```
Comment (matthewh-BQ) [inline guides/cookbook.md]:
```suggestion
		currentType := reader.Type()                             // the first value in the reader is a struct
```

File: guides/cookbook.md [hunk 6]
```diff
@@ -224,6 +224,85 @@ class WriteIonText
...
+		fmt.Println("Current type is:\t" + currentType.String()) // Current type is:   struct
+		reader.StepIn()                                          // step into struct
```
Comment (matthewh-BQ) [inline guides/cookbook.md]:
```suggestion
		reader.StepIn()                                          // step into the struct
```

File: guides/cookbook.md [hunk 7]
```diff
@@ -224,6 +224,85 @@ class WriteIonText
...
+		value, err := reader.StringValue() // retrieve the current value's string value
+		if err != nil {
+			panic("Reading string value failed.")
+		}
+		reader.StepOut()                    // step out of the struct
+		fmt.Println(*fieldName, " ", value) // hello   world
+	}
+}
+```
+Should we have a binary Ion, `NewReaderBytes()` can be used in the same fashion.
```
Comment (byronlin13) [inline guides/cookbook.md]:
```suggestion
If we have Ion in binary format, `NewReaderBytes()` can be used in the same fashion.
```

File: guides/cookbook.md [hunk 8]
```diff
@@ -1632,6 +1906,55 @@ class CsvToIon
...
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	
+	"github.com/amzn/ion-go/ion"
+)
+
+func main() {
+	file, er := os.Open("c://values.cvs")
```
Comment (matthewh-BQ) [inline guides/cookbook.md]: Where does the CVS file come from?

File: guides/cookbook.md [hunk 9]
```diff
@@ -1632,6 +1906,55 @@ class CsvToIon
...
+	if er != nil {
+		panic(er)
+	}
+	defer file.Close()
+
+	buf := strings.Builder{}
+	writer := ion.NewTextWriter(&buf)
+
+	scanner := bufio.NewScanner(file)
+	scanner.Scan() // to skip the first row (header line)
```
Comment (matthewh-BQ) [inline guides/cookbook.md:1936]: Is there context on why we're skipping the first row here?

File: guides/cookbook.md [hunk 10]
```diff
@@ -725,6 +804,52 @@ class PrettyPrint
...
+	err := ion.MarshalTo(writer, l1Val)
+	if err != nil {
+		panic(err)
+	}
+
+	if err := writer.Finish(); err != nil {
+		panic(err)
+	}
+
+	fmt.Println(buf.String())
```
Comment (byronlin13) [inline guides/cookbook.md:848]: Can you comment what this prints

File: guides/cookbook.md [hunk 11]
```diff
@@ -1061,6 +1186,101 @@ class ReadNumericValues
...
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/amzn/ion-go/ion"
+)
+
+func main() {
+	int32Value := 2147483646
+	var int64Value int64 = 9223372036854775807
```
Comment (byronlin13) [inline guides/cookbook.md]:
```suggestion
    int64Value := 9223372036854775807
```

File: guides/cookbook.md [hunk 12]
```diff
@@ -1387,6 +1607,60 @@ class SparseReads
...
 </div>
 
+<div class="tabpane Go" markdown="1">
+
+```Go
+package main
+
+import (
+	"fmt"
+	"github.com/amzn/ion-go/ion"
```
Comment (byronlin13) [inline guides/cookbook.md:1618]: add line to separate

File: guides/cookbook.md (hunk 10 above)
Comment (R-maan) [reply guides/cookbook.md:848]: This is to show the pretty print feature and the output is going to be:
```
   level1: {
     level2: {
       level3: "foo"
     },
     x: 2
   },
   y: [
     a,
     b,
     c
   ]
 }
```
so I cannot put the output as an inline comment.
Also the section where this code is placed, shows the expected output of pretty print

File: guides/cookbook.md (hunk 11 above)
Comment (R-maan) [reply guides/cookbook.md]: I need to explicitly define this value as an int64.

File: guides/cookbook.md (hunk 12 above)
Comment (R-maan) [reply guides/cookbook.md:1618]: done

File: guides/cookbook.md (hunk 8 above)
Comment (R-maan) [reply guides/cookbook.md]: It supposed to be in the directory of where this test is going to be tested. There is explanation about the file in the cookbook and even examples in other languages use the file just like this.

File: guides/cookbook.md (hunk 9 above)
Comment (R-maan) [reply guides/cookbook.md:1936]: The contents of the file are:
```
 id,type,state
 1,foo,false
 2,bar,true
 3,baz,true
```
And is explained in that section.
We are skipping the first line as it is the header and just have the titles

File: guides/cookbook.md (hunk 7 above)
Comment (R-maan) [reply guides/cookbook.md]: I accepted Matthew's already, sorry.

**Pull request closed without merging**